option java_package = "org.godhuli.rhipe";
option java_outer_classname = "REXPProtos";
message REXP {
  enum RClass {
    STRING	= 0;
    RAW		= 1;
    REAL	= 2;
    COMPLEX	= 3;
    INTEGER	= 4;
    LIST	= 5;
    LOGICAL	= 6;
    NULLTYPE	= 7;
//    REAL1       = 8;
//    INTEGER1    = 10;
//    STRING1     = 9;
    ENVIRONMENT = 8;
  }
  enum RBOOLEAN {
    F	= 0;
    T	= 1;
    NA	= 2;
  }

  required RClass   rclass		= 1 ; 
  repeated double   realValue		= 2 [packed=true];
  repeated sint32   intValue		= 3 [packed=true];
  repeated RBOOLEAN booleanValue	= 4;
  repeated STRING   stringValue		= 5;
  optional bytes    rawValue		= 6;
  repeated CMPLX    complexValue	= 7;
  repeated REXP	    rexpValue		= 8;
  repeated string   attrName		= 11;
  repeated REXP     attrValue		= 12;
  repeated ENV      envValue		= 13;
}


message ENV {
	optional string key = 1;
 	optional REXP  value = 2;
}

message STRING {
  optional string strval = 1;
  optional bool isNA = 2 [default=false];
}
message CMPLX {
  optional double real = 1 [default=0];
  required double imag = 2;
}

//message PROMISE {
//  optional REXP environment;
//  optional REXP attributes;
//  optional REXP value;
//  optional REXP expression;
//}
//message CLOSURE {
//  optional attributes
//
//
//
//}